File: graphs/routers.py
def createGraph(connections, n):
    lst = list()
    graph = [lst] * (n + 1)

    for index, connection in enumerate(connections):
        routers = graph[connection] + [index + 2]
        graph[connection] = routers
    
    return graph

paths = [[]]

def dfs(graph, currentVertex, visited):
    visited.append(currentVertex)

    for vertex in graph[currentVertex]:
        if vertex not in visited:
            dfs(graph, vertex, visited.copy())
    paths.append(visited)

def printPath(path):
    for num in path:
        print(num, end=" ")

def main():
    # n = number of routers
    n = int(input())
    # index of the connected routers
    connections = list(map(int, input().split(" ")))
    # print(connections)
    g = createGraph(connections, n)
    visited = []
    dfs(g, 1, visited)

    for path in paths:
        if(len(path) > 0 and path[-1] == n):
            printPath(path)
            print()

if __name__ == "__main__":
    main()

File: graphs/routers.cpp
#include <iostream>
#include <vector>

using namespace std;

struct node {
    int value;
    vector<node *> neighbours;
};

node* checkNode(int &value, node *currNode) {
    if (currNode->value == value) {
        return currNode;
    }

    if (currNode->neighbours.empty()) {
        return nullptr;
    }

    for (int i = 0; i < currNode->neighbours.size(); i++) {
        checkNode(value, currNode->neighbours[i]);
    }

    return nullptr;
}

node* createGraph(vector<int> &v) {
    // first router
    node *root;
    root->value = 1;
    node *currNode = root;

    for (int i = 2; i < v.size() + 1; i++) {
        node *newNode;
        newNode->value = i;

        currNode = checkNode(v[i], currNode);
        if (currNode != nullptr) {
        currNode->neighbours.push_back(newNode);
        }
    }

    return root;
}

int main() {
    // n = number of routers
    int n;
    cin >> n;
    // index of the connection of the routers
    vector<int> connections(n, -1);
    for (int i = 0; i < n - 1; i++) {
        cin >> connections[i];
    }

    node* g = createGraph(connections);

    return 0;
}